java/code/src/org/keyczar/Crypter.java

```diff
@@ -112,61 +111,91 @@ public void decrypt(ByteBuffer input, ByteBuffer output)
-    }
-
-    VerifyingStream verifyStream = cryptStream.getVerifyingStream();
-    if (inputCopy.remaining() < verifyStream.digestSize()) {
-      throw new ShortCiphertextException(inputCopy.remaining());
+    int inputLimit = inputCopy.limit();
+    KeyczarException error = null;
+    
+    boolean collision = keys.size() > 1;
+    
+    for (KeyczarKey key : keys) {
+      error = null;
+      
+      ByteBuffer tempBuffer = output;
+      if (collision) {
+        tempBuffer = ByteBuffer.allocate(output.remaining());
+      }
+      
+      DecryptingStream cryptStream = (DecryptingStream) key.getStream();
+      
+      try {
+        VerifyingStream verifyStream = cryptStream.getVerifyingStream();
+        if (inputCopy.remaining() < verifyStream.digestSize()) {
+          throw new ShortCiphertextException(inputCopy.remaining());
+        }
+
+        // Slice off the signature into another buffer
+        inputCopy.position(inputCopy.limit() - verifyStream.digestSize());
+        ByteBuffer signature = inputCopy.slice();
+
+        // Reset the position of the input to start of the ciphertext
+        inputCopy.reset();
+        inputCopy.limit(inputCopy.limit() - verifyStream.digestSize());
+
+        // Initialize the crypt stream. This may read an IV if any.
+        cryptStream.initDecrypt(inputCopy);
+
+        // Verify the header and IV if any
+        ByteBuffer headerAndIvToVerify = input.asReadOnlyBuffer();
+        headerAndIvToVerify.limit(inputCopy.position());
+        verifyStream.initVerify();
+        verifyStream.updateVerify(headerAndIvToVerify);
+
+        tempBuffer.mark();
+        // This will process large input in chunks, rather than all at once. This
+        // avoids making two passes through memory.
+        while (inputCopy.remaining() > DECRYPT_CHUNK_SIZE) {
+          ByteBuffer ciphertextChunk = inputCopy.slice();
+          ciphertextChunk.limit(DECRYPT_CHUNK_SIZE);
+          cryptStream.updateDecrypt(ciphertextChunk, output);
+          ciphertextChunk.rewind();
+          verifyStream.updateVerify(ciphertextChunk);
+          inputCopy.position(inputCopy.position() + DECRYPT_CHUNK_SIZE);
+        }
+        int lastBlock = inputCopy.position();
+        verifyStream.updateVerify(inputCopy);
+        if (!verifyStream.verify(signature)) {
+          throw new InvalidSignatureException();
+        }
+        inputCopy.position(lastBlock);
+        cryptStream.doFinalDecrypt(inputCopy, tempBuffer);
+        tempBuffer.limit(tempBuffer.position());
+        if (collision) {
+          //Success copy to final output buffer
+          tempBuffer.rewind();
+          output.put(tempBuffer);
+          output.limit(output.position());
+        }
+        return;
+      } catch (KeyczarException e) {
+        LOG.debug(e.getMessage(), e);
+        error = e;
+      } catch (RuntimeException e) {
+        LOG.debug(e.getMessage(), e);
+        error = new InvalidSignatureException();
```

Comment:
Since a runtime exception is unknown and unfortunately possible, I think it's safer not to propagate them out. InvalidSignature seems appropriate from both a security perspective if it goes unhandled in the consuming application and logical if it does get handled in the consuming application.